test/kernels/test_qkernel.py

```diff
@@ -358,5 +359,73 @@ def test_ydim(self):
+
+    def setUp(self):
+        super().setUp()
+
+        # Create an arbitrary 2-qubit feature map circuit
+        circ1 = ZZFeatureMap(2)
+        circ2 = ZZFeatureMap(2)
+        user_params = circ2.parameters
+        for i, _ in enumerate(user_params):
+            user_params[i]._name = f"θ[{i}]"
+
+        self.feature_map = circ1.compose(circ2).compose(circ1)
+        self.user_parameters = user_params
+
+    def test_bad_param_values(self):
+        """Test handling of bad parameter assignments."""
+        # Instantiate a QuantumKernel but do not set the user_parameters field
+        qkclass = QuantumKernel(feature_map=self.feature_map)
+
+        # Try to set the user parameters using only values and ensure an
+        # error is raised.
+        user_param_values = [np.pi / 2, np.pi / 4]
+        with self.assertRaises(ValueError):
+            qkclass.assign_user_parameters(user_param_values)
+
+    def test_user_parameters(self):
+        """Test assigning user parameters"""
+        # Ensure we can instantiate a QuantumKernel with user parameters
+        qkclass = QuantumKernel(feature_map=self.feature_map, user_parameters=self.user_parameters)
+        self.assertEqual(qkclass.user_parameters, self.user_parameters)
+
+        # Test iterative user parameter binding
+        qkclass.assign_user_parameters({qkclass.user_parameters[0]: np.pi / 2})
+        self.assertEqual(qkclass.feature_map.num_parameters, 3)
+        qkclass.assign_user_parameters({qkclass.user_parameters[1]: np.pi / 2})
+        self.assertEqual(qkclass.feature_map.num_parameters, 2)
+
+        # Instantiate a QuantumKernel without specifying user parameters
+        qkclass = QuantumKernel(feature_map=self.feature_map)
+
+        # Bind user parameters to real values
+        user_param_values = [np.pi / 3, np.pi / 6]
+        param_binds = dict(zip(self.user_parameters, user_param_values))
+        qkclass.assign_user_parameters(param_binds)
+
+        # Ensure the two user parameters were bound correctly
+        bound_user_param_vals = [
+            val for val in qkclass.user_param_binds.values() if isinstance(val, numbers.Number)
+        ]
+        self.assertEqual(user_param_values, bound_user_param_vals)
+
+        # Assign params to some new values, and also test the bind_user_parameters interface
+        user_param_values = [np.pi / 4, np.pi / 2]
+        param_binds = dict(zip(self.user_parameters, user_param_values))
+        qkclass.bind_user_parameters(param_binds)
+
+        # Ensure the old values are properly overwritten in the feature map
+        bound_user_param_vals = [
+            val for val in qkclass.user_param_binds.values() if isinstance(val, numbers.Number)
+        ]
+        self.assertEqual(user_param_values, bound_user_param_vals)
+
+        # Ensure unbound feature map still holds all parameters of input feature map
+        self.assertEqual(len(qkclass.unbound_feature_map.parameters), 4)
```

Comment:
This test seems to comprise a series of tests. If an assert fails then the rest of the test will not be run. If the test aspects are independent, which at first glance they seem to be, then it is better to use the subTest capability (assuming that its not perhaps better to have individual tests for each part). Each subTest will be run even if one fails - idea being you get a fuller picture of what is/is not working then if the test fails at the first bit. See this for an example https://github.com/Qiskit/qiskit-machine-learning/blob/223c4127cf8874afeb349eba86d6bca1f498d14f/test/circuit/library/test_raw_feature_vector.py#L35-L49